### PR TITLE
Fix revealed default config in header if requirements in subfolder

### LINKED ISSUE #1903: pip-compile reveals default `--config=pyproject.toml` option in header
There is a bug with pip 6.14.0. I just use `pip-compile --allow-unsafe --generate-hashes --output-file=requirements/test.txt --resolver=backtracking --strip-extras requirements/test.in`
but now it adds `--config=pyproject.toml` even there is no configuration in pyproject.toml
When running pip-compile click context has no src_files, so select_config_file returns pyproject.toml, but after compiling click context has src_files as requirements/test.in, so select_config_file returns None and so pyproject.toml != None

Making `--config` is_eager=False works

_Originally posted by @q0w in https://github.com/jazzband/pip-tools/issues/1893#issuecomment-1624405135_

### PR DESCRIPTION
**Pull request by atugushev**
Ensure `src_files` are handled prior to `--config`, as `override_defaults_from_config_file` is dependent on `src_files`.

Fixes #1903.

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [x] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by atugushev** [comment]
@q0w this should fix your issue too #1902.

**Comment by q0w** [comment]
what about this. I think it should pickup config from pyproject.toml
https://github.com/jazzband/pip-tools/issues/1902#issuecomment-1624429055

File: tests/test_utils.py [hunk 1]
```diff
@@ -410,6 +410,29 @@ def test_get_compile_command_with_config(tmpdir_cwd, config_file, expected_comma
...
+    """
+    default_config_file = Path("pyproject.toml")
+    default_config_file.touch()
+
+    (tmpdir_cwd / "subdir").mkdir()
+    req_file = Path("subdir/requirements.in")
+    req_file.touch()
+
+    with open(req_file, "w"):
+        pass
```
Comment (webknjaz) [inline tests/test_utils.py:429]: Wouldn't this be simpler?
```suggestion
    req_file.write_bytes(b"")
```

File: tests/test_utils.py [hunk 2]
```diff
@@ -410,6 +410,29 @@ def test_get_compile_command_with_config(tmpdir_cwd, config_file, expected_comma
...
+def test_get_compile_command_does_not_include_default_config_if_reqs_file_in_subdir(
+    tmpdir_cwd,
+):
+    """
+    Test that ``get_compile_command`` does not include default config file
+    if requirements file is in a subdirectory.
+    Regression test for issue GH-1903.
+    """
+    default_config_file = Path("pyproject.toml")
+    default_config_file.touch()
```
Comment (webknjaz) [inline tests/test_utils.py:422]: Can we also test the case when the `[tool.pip-tools]` section is present and empty?

Also, how about a `.pip-tools.toml` with an empty or missing `[tool.pip-tools]` and `pyproject.toml` with a non-empty one? I think that the `pyproject.toml` shouldn't be taken into account in this case.

Comment (atugushev) [reply tests/test_utils.py:422]: Okay, I'll add those combinations to parametrized tests 👍🏻

File: tests/test_utils.py (hunk 1 above)
Comment (atugushev) [reply tests/test_utils.py]: Yes, indeed. Applied.

File: tests/test_utils.py (hunk 2 above)
Comment (atugushev) [reply tests/test_utils.py]: Addressed in 64445d00524f2b76af6825c2f79424b179d914bc

Comment (webknjaz) [reply tests/test_utils.py]: I was hoping to see what happens if both files exist too..

Comment (atugushev) [reply tests/test_utils.py]: This looks like out of scope of `get_compile_command`. We can add more unit tests to `select_config_file()`.

Comment (webknjaz) [reply tests/test_utils.py]: If it's out of the scope, it'd be good to add the tests separately.

Comment (atugushev) [reply tests/test_utils.py]: Addressed In 783a6e0cb37ed52c2cdf07a88c5a5e5bdfca8d06

**Comment by chrysle** [comment]
I guess #1906 is to be merged first?

**Comment by atugushev** [comment]
Looks like the bug is fixed now. Ready for another round of review.

**Comment by atugushev** [comment]
FTR, resolved some conflicts after merging with the main branch.

File: tests/conftest.py [hunk 3]
```diff
@@ -452,13 +452,16 @@ def make_config_file(tmpdir_cwd):
...
+        pyproject_param: str,
+        new_default: Any,
+        config_file_name: str = DEFAULT_CONFIG_FILE_NAMES[0],
     ) -> Path:
-        # Make a config file with this one config default override
-        config_path = tmpdir_cwd / pyproject_param
-        config_file = config_path / config_file_name
-        config_path.mkdir(exist_ok=True)
+        # Create a nested directory structure if config_file_name includes directories
+        config_dir = Path(tmpdir_cwd / config_file_name).parent
```
Comment (theryanwalker) [inline tests/conftest.py:460]:
```suggestion
        config_dir = (tmpdir_cwd / config_file_name).parent
```

The `tmpdir_cwd` already yields a Path-like object, so it's unnecessary.

**Review by theryanwalker — APPROVED** [review]
Left a minor comment. Otherwise, LGTM!

File: tests/conftest.py (hunk 3 above)
Comment (atugushev) [reply tests/conftest.py]: Ah, no. Can', because `as_posix()` shows full path and breaks tests.

**Comment by atugushev** [comment]
Thanks @webknjaz and @theryanwalker for the review!

**Pull request merged**